### PR TITLE
When rescheduling containers, don't try to create them with multiple networks

### PR DESCRIPTION
**Pull request by wsong**
The Docker daemon does not support creating containers with multiple network
configurations, so we need to reconnect them one at a time. This was already
being done, but we were not clearing out the Config's NetworkSettings object
so the daemon was still returning an error.

I also cleared out the IP address for EndpointsSettings objects for networks
that did not have subnets, because setting an explicit IP address for
networks without subnets also results in an error from the daemon.

Signed-off-by: Wayne Song <wsong@docker.com>

### CODE REVIEW

File: cluster/watchdog.go [hunk 1]
```diff
@@ -107,7 +107,7 @@ func (w *Watchdog) rescheduleContainers(e *Engine) {
 			clusterNetworks := w.cluster.Networks().Uniq()
 			for networkName, endpoint := range c.Info.NetworkSettings.Networks {
 				net := clusterNetworks.Get(endpoint.NetworkID)
-				if net != nil && net.Scope == "global" {
+				if net != nil && (net.Scope == "global" || net.Scope == "swarm") {
```
Comment (nishanttotla) [inline cluster/watchdog.go]: Why is `net.Scope == "swarm"` required here? Isn't this something we were discussing in another PR as well?

File: cluster/watchdog.go [hunk 2]
```diff
@@ -121,6 +121,9 @@ func (w *Watchdog) rescheduleContainers(e *Engine) {
 			}
 		}
 
+		// Clear out the network configs, since we cannot create a
+		// container with multiple networks.
+		c.Config.NetworkingConfig.EndpointsConfig = nil
```
Comment (nishanttotla) [inline cluster/watchdog.go]: One of the unfortunate things about this piece of code is that some information, particularly endpoints are replicated in multiple locations. Make doubly sure that this works on testing. ie `c.Config.NetworkingConfig.EndpointsConfig` is the data structure you want to change.

File: cluster/watchdog.go (hunk 1 above)
Comment (wsong) [inline cluster/watchdog.go]: When running Swarm with Swarm mode enabled, we need to disconnect and reconnect swarm-scoped networks as well, not just global-scoped ones. swarm-scoped networks are, in effect, globally-scoped ones.

File: cluster/watchdog.go (hunk 2 above)
Comment (wsong) [inline cluster/watchdog.go]: It worked when I tested this manually on a 1.13 cluster.

Comment (dongluochen) [inline cluster/watchdog.go]: When the network configs is wiped, `newContainer` wouldn't get this config. The re-created `newContainer` would have a different `NetworkingConfig.EndpointsConfig`. I suspect `newContainer` should restore the original's `Config.NetworkingConfig.EndpointsConfig`. Can you take a look?

Comment (wsong) [inline cluster/watchdog.go]: `newContainer` is not after this function returns; we just use it to get the new container's ID and engine object.

Comment (dongluochen) [inline cluster/watchdog.go]: The created container would not have the original's `NetworkingConfig.EndpointsConfig`, isn't it?

Comment (wsong) [inline cluster/watchdog.go]: That's correct, but it'll get updated when we do the `NetworkConnect` calls. The reason why we can't create the new container with the `EndpointsConfig` object is that you can't create a container that's attached to multiple networks.

Comment (dongluochen) [inline cluster/watchdog.go]: I think `NetworkConnect` would add the `EndpointsCofnig` for the network to be connected. Are there any other config that's not **global**(or **swarm**)?

Comment (wsong) [inline cluster/watchdog.go]: It appears the only other type of scope is `local`:

https://github.com/docker/docker/blob/5497f4758d14f6f65c722838d342f634e353b08e/api/server/router/network/network_routes.go#L182
https://github.com/docker/libnetwork/blob/82905ed5312c9b3403e5266aab5c2b2dd346ce5a/datastore/datastore.go#L113

Comment (dongluochen) [inline cluster/watchdog.go]: We need to preserve the `local` type then.

Comment (dongluochen) [inline cluster/watchdog.go]: I mean, the `local` type configuration should be used when creating `newContainer`.

Comment (wsong) [inline cluster/watchdog.go]: We weren't disconnecting/reconnecting networks for local networks before. In fact, doing so for the `local` networks is almost guaranteed to fail since we're rescheduling this container on a new node, which will not have the local networks the container previously relied upon.

Comment (dongluochen) [inline cluster/watchdog.go]: Let's say the original container has a host TCP port 2388, and is connected to an overlay network `ovnet`. With this change, the original host port is not created on `newContainer`, right?

Comment (wsong) [inline cluster/watchdog.go]: Host ports are stored in `HostConfig.PortBindings`, not in the `EndpointsConfig` map.

Comment (dongluochen) [inline cluster/watchdog.go]: So what goes into `local` in `Config.NetworkingConfig.EndpointsConfig`?

Comment (wsong) [inline cluster/watchdog.go]: Bridge networks can have their scope field set to `local`, indicating that they only exist on a single node.

Comment (dongluochen) [inline cluster/watchdog.go]: That's another trouble in container rescheduling. Should we check if this container has bridge (local) network setup and refuse to reschedule it because it's node dependent, or ignore bridge network?

Comment (wsong) [inline cluster/watchdog.go]: Prior to this change, we essentially ignored the trouble; since we attempted to preserve the network configuration, if the container was connected to a bridge network, we'd try to recreate it on a new node with that same network config. If there was a network with the same name, it'd work; otherwise, it'd fail.

I can keep around networks that do not have the `swarm` or `global` scope if you'd like this change to have as little impact as possible.

Comment (dongluochen) [inline cluster/watchdog.go]: > keep around networks that do not have the swarm or global scope

I prefer that. After all there is default bridge network named `bridge`. And users may create same bridge network on every node.

Comment (wsong) [inline cluster/watchdog.go]: Done.

File: cluster/watchdog.go [hunk 3]
```diff
@@ -121,6 +121,18 @@ func (w *Watchdog) rescheduleContainers(e *Engine) {
 			}
 		}
 
+		// Clear out the network configs that we're going to reattach
+		// later.
+		endpointsConfig := map[string]*network.EndpointSettings{}
+		for k, v := range c.Config.NetworkingConfig.EndpointsConfig {
+			_, nameExists := globalNetworks[v.Name]
```
Comment (dongluochen) [inline cluster/watchdog.go]:
```
13.56s$ GOOS=darwin go build
# github.com/docker/swarm/cluster
cluster/watchdog.go:128: v.Name undefined (type *network.EndpointSettings has no field or method Name)
cluster/watchdog.go:129: v.ID undefined (type *network.EndpointSettings has no field or method ID)
The command "GOOS=darwin go build" exited with 2.
```

Comment (wsong) [inline cluster/watchdog.go]: Fixed.

**Comment by dongluochen** [comment]
@wsong Test failure on engine-1.9. PTAL. 

```
# [signal SIGSEGV: segmentation violation code=0x1 addr=0x40 pc=0x61228a]
# 
# goroutine 406 [running]:
# panic(0xc17ae0, 0xc4200120a0)
# 	/usr/local/go/src/runtime/panic.go:500 +0x1a1
# github.com/docker/swarm/cluster.(*Watchdog).rescheduleContainers(0xc4201f10c0, 0xc4204202c0)
# 	/go/src/github.com/docker/swarm/cluster/watchdog.go:129 +0x55a
# created by github.com/docker/swarm/cluster.(*Watchdog).Handle
# 	/go/src/github.com/docker/swarm/cluster/watchdog.go:29 +0x12f
```

**Comment by wsong** [comment]
Whoops, forgot some parens. Fixed.

**Comment by dongluochen** [comment]
I'm a little puzzled by this in [engine-1.10 test](https://jenkins.dockerproject.org/job/Swarm-PRs%20(engine%201.10)/1789/console). Don't have time to dig into it now. 

```
# time="2017-02-01T22:14:49Z" level=info msg="New leader elected: 127.0.0.1:6702" 
# 2017/02/01 22:14:49 http: panic serving 127.0.0.1:41876: runtime error: invalid memory address or nil pointer dereference
# goroutine 30 [running]:
# net/http.(*conn).serve.func1(0xc420123400)
# 	/usr/local/go/src/net/http/server.go:1491 +0x12a
# panic(0xc17ae0, 0xc4200120a0)
# 	/usr/local/go/src/runtime/panic.go:458 +0x243
# github.com/docker/swarm/api.(*Replica).ServeHTTP(0xc420053a90, 0x110a1e0, 0xc420126410, 0xc4201345a0)
# 	/go/src/github.com/docker/swarm/api/replica.go:38 +0x277
# github.com/docker/swarm/api.(*dispatcher).ServeHTTP(0xc420310c40, 0x110a1e0, 0xc420126410, 0xc4201345a0)
# 	/go/src/github.com/docker/swarm/api/server.go:33 +0x5a
# net/http.serverHandler.ServeHTTP(0xc420123100, 0x110a1e0, 0xc420126410, 0xc4201345a0)
# 	/usr/local/go/src/net/http/server.go:2202 +0x7d
# net/http.(*conn).serve(0xc420123400, 0x110b0e0, 0xc4203ee740)
# 	/usr/local/go/src/net/http/server.go:1579 +0x4b7
# created by net/http.(*Server).Serve
# 	/usr/local/go/src/net/http/server.go:2293 +0x44d
```

**Comment by nishanttotla** [comment]
@dongluochen you're right, this is actually a little concerning. Let me look into it. It's not related to this PR though, so we can restart CI one more time and investigate the issue in parallel.

**Comment by wsong** [comment]
Looks like that failure occurred again, but I'm quite baffled as to how it's possible. The line it's failing on is:

https://github.com/docker/swarm/blob/master/api/replica.go#L38

```
p.handler.ServeHTTP(w, r)
```

But there's no way that `p.handler` could ever be nil. `Replica` objects are always instantiated with a non-nil `handler`. I suppose the receiver of that method (i.e. `p`) could be nil as well, but again, I don't see how; we should never reach a code path where either of those variables could be nil.

**Comment by nishanttotla** [comment]
Uh oh, this is quite concerning. I agree that I don't see how we could have an exception on that line. I couldn't reproduce locally :-/

**Comment by dongluochen** [comment]
I suspect it's related to leader change. The replica's data structure might change because of replica update, or indirectly underlayer connection change. May need to look into that. I think it's not related to this change. 

```
# time="2017-02-01T22:14:49Z" level=info msg="New leader elected: 127.0.0.1:6702" 
```

**Comment by dongluochen** [comment]
I opened #2622 to track the panic. I don't see it related to this PR. 

@wsong Can you clean up the debug message in api/replica.go?

**Comment by wsong** [comment]
Removed debug message.

File: api/replica.go [hunk 4]
```diff
@@ -5,6 +5,8 @@ import (
 	"fmt"
 	"net/http"
 	"strings"
+
+	log "github.com/Sirupsen/logrus"
```
Comment (dongluochen) [inline api/replica.go]: Please revert the change to this file.

Comment (wsong) [inline api/replica.go]: Whoops, done.

**Comment by dongluochen** [comment]
LGTM
